### PR TITLE
property: do not overwrite the NULL-provider cache entry on set

### PR DESCRIPTION
**Pull request by bukka**
ossl_method_store_cache_set inserts two entries per method: one keyed on (nid, prop_query, prov) and one keyed on (nid, prop_query) with a NULL provider, used to match "any provider" lookups.

Previously the set path always replaced the NULL-provider entry. When a second provider cached the same nid, its method became the result for "any provider" lookups, even though an earlier provider was already cached. A shared nid could then resolve to the wrong provider: a certificate's SPKI would decode through that provider's keymgmt and X509_check_private_key would fail with a key value mismatch.

Only insert the NULL-provider entry when one does not already exist, so the first provider to cache the nid owns it, matching the order ossl_method_store_fetch would select.

Assisted-by: Claude:claude-opus-4-8

### CODE REVIEW

**Comment by nhorman** [comment]
I think I see what you're saying:
1) The default provider adds a method  with the tuple <nid=1, prop_query=<foo>, provider=<p1>>
2) Another provider adds a method with the tuple <nid=1, prop_query=<foo>, provider=<p2>>

In that scenario (2) would replace the <nid=1,prop_query=<foo>,provider=NULL>> entry that (1) added with the version that (2) added.

do I have that right?

If so, I have some questions:
1) Where are we adding two algorithms with the same nid and property query?  ostensibly the property query has to be the same here, or they wouldn't be findable, no?  It seems odd that we would do so, as not having that differentiation means that the algorithm you get from a fetch operation will be at the mercy of openssl, even if you supply a specific property_query during the fetch.

2) @mattcaswell warned me about hyrums law [here](https://github.com/openssl/openssl/pull/31018#issuecomment-4708210600), and this seems to have fallen afoul of it, but it gives me some concern that the pkcs11 provider (or some openssl function its calling), is making assumptions about what will get returned from `EVP_*_fetch`.  If you call one of the EVP fetch functions without a specific property query, there is no guarantee that you will get the same method from the same provider on subsequent calls.  Is there a corresponding change in pkcs11_provider that needs to be made to expressly ask for the pcks11_provider method for an operation to avoid hitting this in the future if other method store changes are made that impact the method selection behavior.

**Review by nhorman — APPROVED** [review]

**Comment by bob-beck** [comment]
I mean "maybe" - but since there are no guarantees here, and to revert back to hyrams law, If this is happening today, Might someone not be depending on this property to get the result they want from loading another provider? and that makes this a breaking change for those folks. 

It's not about what you "meant" for it to do, it's about what it is actually doing, and what, by extension, people using this will have managed to figure out and adapt around this behaviour to do what they want. 

So sure, I agree in hindsight this probably makes sense, but is this something that needs to wait for a major?

**Review by bob-beck — CHANGES_REQUESTED** [review]
I'm concerned this could be a breaking change for folks loading providers that expect it to behave like this "because it did" and have worked around / adapted to this behaviour.

**Comment by bob-beck** [comment]
sounds to me like pkcs11 provider should use a different propquery for registering different providers if they want to distinguish here. In theory, if I register an implementation for the same nid matching the same propquery - it's valid for me to get that implementation if I am querying for any provider, and I should not care which provider gives me an
implementation that matches.

Right?

If I want to distinguish further, I need to differentiate by asking for a particular provider or a more specific propquery.

**Comment by nhorman** [comment]
I don't think it needs to wait for a major.  The change that altered the behavior just went in last week, so we've not had a release on which people will begin to use/rely on the new behavior.  this restores the previous behavior "in time", so I think we're good on that account.

I think the bigger questions here are:
1) How exactly did this situation arise?  It really seems to me like, for this to happen, in such a way that this change fixes it, we must have two providers registering the same nid with the same property query, which seems ill advised, as it means the EVP fetch apis have no way to differentiate those algorithms save for a provider specific fetch operation.

2) how much we want to bend to hyrum here.  I get that this api has behaved in such a way that people have come to have explicit expectations of it, and given that this behavior is a recent change, and that we can restore the old behavior fairly easily in this case, it seems like a fine fix to take.  That said though, making this change seems to imply to me that we are guaranteeing a lookup behavior despite the fact that we make no guarantees that the same algorithm implementation will be returned from successive calls to EVP fetch apis (considering the case here in which you do the following):
```
OSSL_PROVIDER_load(foo)
EVP_MD_fetch(...,"SHA256", NULL)
OSSL_PROVIDER_load(bar)
EVP_MD_fetch(...,"SHA256", NULL)
```
Its a contrived example that showcases the inverse problem here, but if you want to ensure that the second call to EVP_MD_fetch gives you the implementation from the bar provider, you need to call:
EVP_MD_fetch(..., "SHA256", "provider=bar");

More generally, any time an application needs to guarantee that a specific provider implementation is returned, then the call to the fetch operation needs to specify that in the property query.

Again, I'm ok restoring the old behavior with this PR, as it solves the problem before the new behavior has been released, but I think its worth the exercise to understand where the pkcs11 tls test made the assumption that it would always get the implementation from a specific provider.

**Review by simo5 — APPROVED** [review]
LGTM

**Comment by simo5** [comment]
> I'm concerned this could be a breaking change for folks loading providers that expect it to behave like this "because it did" and have worked around / adapted to this behaviour.

This is a recent change that broke pkcs11-provider, so it is the opposite, and it is a regression.

**Comment by simo5** [comment]
> 1. How exactly did this situation arise?  It really seems to me like, for this to happen, in such a way that this change fixes it, we must have two providers registering the same nid with the same property query, which seems ill advised, as it means the EVP fetch apis have no way to differentiate those algorithms save for a provider specific fetch operation.

This is part of the design of providers, any provider can offer a specific algorithm, there is no need to have a special property query on it, the FIPS provider also offers the same algorithms as the default provider ...

In particular requiring properties would break the transparent handling of operations when they are keyed on the used private Key, which can be sourced via store operations from a specific provider, in which case we need to use that provider to implement the operation.

Now, this does not apply to digests, but those are kind of a special case, we have config options in the pkcs11-provider to mask off all digest operations.

The good thing is that the pkcs11-provider exercises all thee corners of the providers API that are completely masked in other openssl tests because in most cases when using default + FIPS provider the fact they are basically the same implementation hides a lot of issues that can otherwise happen.

**Comment by bob-beck** [comment]
> I don't think it needs to wait for a major. The change that altered the behavior just went in last week, so we've not had a release on which people will begin to use/rely on the new behavior. this restores the previous behavior "in time", so I think we're good on that account.
> 
> I think the bigger questions here are:
> 
> 1. How exactly did this situation arise?  It really seems to me like, for this to happen, in such a way that this change fixes it, we must have two providers registering the same nid with the same property query, which seems ill advised, as it means the EVP fetch apis have no way to differentiate those algorithms save for a provider specific fetch operation.
> 2. how much we want to bend to hyrum here.  I get that this api has behaved in such a way that people have come to have explicit expectations of it, and given that this behavior is a recent change, and that we can restore the old behavior fairly easily in this case, it seems like a fine fix to take.  That said though, making this change seems to imply to me that we are guaranteeing a lookup behavior despite the fact that we make no guarantees that the same algorithm implementation will be returned from successive calls to EVP fetch apis (considering the case here in which you do the following):
> 
> ```
> OSSL_PROVIDER_load(foo)
> EVP_MD_fetch(...,"SHA256", NULL)
> OSSL_PROVIDER_load(bar)
> EVP_MD_fetch(...,"SHA256", NULL)
> ```
> 
> Its a contrived example that showcases the inverse problem here, but if you want to ensure that the second call to EVP_MD_fetch gives you the implementation from the bar provider, you need to call: EVP_MD_fetch(..., "SHA256", "provider=bar");
> 
> More generally, any time an application needs to guarantee that a specific provider implementation is returned, then the call to the fetch operation needs to specify that in the property query.
> 
> Again, I'm ok restoring the old behavior with this PR, as it solves the problem before the new behavior has been released, but I think its worth the exercise to understand where the pkcs11 tls test made the assumption that it would always get the implementation from a specific provider.

ok, that finally makes sense to me, in that case I'm good with it

**Review by bob-beck — APPROVED** [review]

**Comment by bukka** [comment]
> I think I see what you're saying:
> 
> 1. The default provider adds a method  with the tuple <nid=1, prop_query=, provider=>
> 2. Another provider adds a method with the tuple <nid=1, prop_query=, provider=>
> 
> In that scenario (2) would replace the <nid=1,prop_query=,provider=NULL>> entry that (1) added with the version that (2) added.
> 
> do I have that right?
> 

Yes, this is exactly it. This is only about order of what cached value is stored. As I understand it, previously it had this collision check and effectively ignored the the follow up cache write for the same nid (and empty prop_query and provider) - it means first cache write wins. With your change, it got changed that effectively the follow up cache write overwrote the old one. This effectively just restores the previous behaviour.

> 1. Where are we adding two algorithms with the same nid and property query?  ostensibly the property query has to be the same here, or they wouldn't be findable, no?  It seems odd that we would do so, as not having that differentiation means that the algorithm you get from a fetch operation will be at the mercy of openssl, even if you supply a specific property_query during the fetch.

Think this was well answered by @simo5 .

> 2. @mattcaswell warned me about hyrums law [here](https://github.com/openssl/openssl/pull/31018#issuecomment-4708210600), and this seems to have fallen afoul of it, but it gives me some concern that the pkcs11 provider (or some openssl function its calling), is making assumptions about what will get returned from `EVP_*_fetch`.  If you call one of the EVP fetch functions without a specific property query, there is no guarantee that you will get the same method from the same provider on subsequent calls.  Is there a corresponding change in pkcs11_provider that needs to be made to expressly ask for the pcks11_provider method for an operation to avoid hitting this in the future if other method store changes are made that impact the method selection behavior.

I guess this comes back again to the fact that there was some deterministic behaviour (first cache wins in this case) and if it gets changed, there will be always some risk of a break. I can't really think about anything that pkcs11-provider could do to address (other than explicitly setting propquery which doesn't seem like the right solution). Maybe it could improve getting the attrs so it sort of matches. Interestingly there was some of this logic dropped as dead code recently: https://github.com/openssl-projects/pkcs11-provider/pull/734/changes . But I reverted it and tested it with that code back and it didn't really work out so there might be some missing parts.

**Comment by nhorman** [comment]
@simo5 

> In particular requiring properties would break the transparent handling of operations when they are keyed on the used private Key, which can be sourced via store operations from a specific provider, in which case we need to use that provider to implement the operation.
> 

This is true, if the caller is ok with getting any implementation.  If (as you said previously) you have two offered algorithms, calling an EVP fetch operation without a distinguishing property query means you are not guaranteed to get any specific algorithm.  The only reason you might is by relying on behavior of the cache that has never been guaranteed.

We have the opportunity to restore the behavior here, and its not hard to do, so I'm ok with that, but I don't think its right to assume the implicit behavior should be relied upon in perpetuity.  It seems clear from this bug, that, somewhere in the pkcs11 tls  test, an assumption was made that an EVP fetch would return an algorithm from the pkcs11 provider, and openssl returned an algorithm from the default (or other) provider instead, causing this problem.  

We're currently fixing that by restoring the old behavior which happens to guarantee a consistent algorithm source, but pkcs11 shouldn't rely on that.  We should identify where that fetch happened, and if its not specifically requesting an alg from pkcs11, then it likely should be.

> Now, this does not apply to digests, but those are kind of a special case, we have config options in the pkcs11-provider to mask off all digest operations.
> 
> The good thing is that the pkcs11-provider exercises all thee corners of the providers API that are completely masked in other openssl tests because in most cases when using default + FIPS provider the fact they are basically the same implementation hides a lot of issues that can otherwise happen.

**Comment by simo5** [comment]
> We're currently fixing that by restoring the old behavior which happens to guarantee a consistent algorithm source, but pkcs11 shouldn't rely on that. We should identify where that fetch happened, and if its not specifically requesting an alg from pkcs11, then it likely should be.

I would definitely love to know more about this and why it fails.
I was currently tracking the failure here: https://github.com/openssl-projects/pkcs11-provider/issues/742 but those logs have nothing to do with context duplication, so the error must be deep somewhere (thanks @bukka for the gdb session) and we get to those logs only as a side effect somewhow.

I guess we'll keep that ticket open to see if we can track this more deeply, although once this PR lands the reproducer will be gone too.

**Comment by nhorman** [comment]
@simo5  @bukka 
FWIW, to illustrate the point I was trying to make, this patch also fixes the problem with the tls test in pkcs11_provider, and I think explains what I was trying to get at.

The method store cache definitely has (as you identified @bukka) a change in behavior in which the same algorithm won't get returned if the cache is updated during some load operation.  However the three points below in the tlssetkey test and some of its utilities implicitly rely on the old behavior by not providing a property query string to select the provider they want to get.

```
diff --git a/tests/tlssetkey.c b/tests/tlssetkey.c
index 3fef8a7..679ec74 100644
--- a/tests/tlssetkey.c
+++ b/tests/tlssetkey.c
@@ -21,9 +21,9 @@ int main(int argc, char *argv[])
         exit(EXIT_FAILURE);
     }
     cert = load_cert(argv[1], NULL, NULL);
-    pkey = load_key(argv[2]);
+    pkey = load_key_ex(argv[2], "?provider=pkcs11");
 
-    ctx = SSL_CTX_new(TLS_client_method());
+    ctx = SSL_CTX_new_ex(NULL, "?provider=pkcs11", TLS_client_method());
     if (!ctx) {
         fprintf(stderr, "Failed to create SSL Context\n");
         ossl_err_print();
diff --git a/tests/util.c b/tests/util.c
index 7b9eccb..8a5e025 100644
--- a/tests/util.c
+++ b/tests/util.c
@@ -117,7 +117,7 @@ X509 *load_cert(const char *uri, const UI_METHOD *ui_method, void *ui_data)
         exit(EXIT_FAILURE);
     }
 
-    store = OSSL_STORE_open(uri, ui_method, ui_data, NULL, NULL);
+    store = OSSL_STORE_open_ex(uri, NULL, "?provider=pkcs11", ui_method, ui_data, NULL, NULL, NULL);
     if (store == NULL) {
         fprintf(stderr, "Failed to open store: %s\n", uri);
         ossl_err_print();
```

By providing this property query "?provider=pkcs11" (i.e. "give me the pkcs11 providers alg if you can, otherwise anything will do"), we can guarantee that, for any alg we ask for, we're going to get the pkcs11 implementation if its available.

I'm not saying that this is the better fix, I'm only trying to illustrate my point about relying on the alg returned from the various fetch operations.

Without the property query in place here, SSL_CTX_new and OSSL_STORE_open fetch whatever algs matche first, and thats at the mercy of the implementation. 

We're in agreement that restoring the old behavior is the most expedient path to resolve this issue, as its pretty straightforward to do right now, but that might change in the future (it might not, but its not a defined property of the cache to consistently return the same algorithm for any period of time, and any small adjustment to the method store cache, may impact this).

And so, if you need to guarantee that the algs you use come from a specific provider, this may be good future proofing.

**Comment by simo5** [comment]
@nhorman that patch is yet another way to cover up the problem.

The use of provider=XX in pkcs11-provider test is carefully controlled to make sure we test the "default" case vs the "force pkcs11-provider" case.

When you know you need to fetch an algorithm for a a specific provider of course you'd want to use a propquery in an application. But remember that a crapton of applications just do not do it, so our tests are carefully crafted to make sure applications can be made to use pkcs11-provider hosted keys w/o having to modify applications (as long as those applications use either the STORE API via URIs or the special PEM format via openssl decoders).

**Comment by simo5** [comment]
that said given digests are not using keys, I may be convinced to make them available only when specifying provider=pkcs11, but that is only a subset, all other caching still has to work consistently and not flap around. First winner take all works because it is consistent even though sometimes you may end up using an undesirable provider (rare). But flapping as the code does now just breaks things, so this patch is really needed and the right thing to do IMO.

**Comment by bukka** [comment]
@nhorman I get your point and I think it makes sense. Providers should make sure that they select the right provider for the operation. That's what the new fix from @simo5 does: https://github.com/openssl-projects/pkcs11-provider/pull/744 and it fixes the problem too. Even though this is the right approach, I think it will be extremly hard to follow. The authors will just have to keep this in their mind when they use any OpenSSL call that does fetching and relies on certain provider which I don't think is something that will be followed. So I don't think it's just something that we can break and we should make sure that this is tested. I will add to my list of testing things to do so there are tests for this behavior and possibly other edge cases.

**Comment by nhorman** [comment]
@bukka I agree.  We should definitely restore the old behavior here just to avoid all the potential trouble with apps that expect the old cache behavior, but we should take the opportunity to future proof pkcs11 against subsequent openssl behavioral cahanges in this area, and @simo5  pr seems to cover that.

Thank you!

**Comment by simo5** [comment]
@nhorman I think we have enough approvals to merge this if you have cycles :)

**Comment by nhorman** [comment]
@simo5 we do, I'm just waiting on the 24 hour timer to expire (ETA is 1:33pm EDT).  I'll merge it as soon as that happens.

**Comment by nhorman** [comment]
merged to master, thank you.

**Comment by simo5** [comment]
thanks @nhorman I tend to forget the 24h cooldown thing ...

**Comment by bukka** [comment]
The follow up test for this in https://github.com/openssl/openssl/pull/31738
